Mathlib/Tactic/Find2.lean

```diff
@@ -0,0 +1,352 @@
+      NameSet.union (m₁.findD needle {}) (m₂.findD needle {})
+
+    -- Filter by name patterns
+    let hits2 := hits.toArray.filter fun n => args.name_pats.all fun p =>
+      p.isInfixOf n.toString
+
+    -- Filter by term patterns
+    let hits3 <- hits2.filterM fun n => do
+      let env <- getEnv
+      if let some ci := env.find? n then do pats.allM (· ci)
+      else return false
+
+    -- Sort to bring lemmas with small names to the top, and keep related lemmas together
+    let hits4 := hits3.qsort Name.lt
+
+    let summary ← IO.mkRef MessageData.nil
+    let add_line line := do summary.set $ (← summary.get) ++ line ++ Format.line
+
+    let needles_list := MessageData.andList (needles.toArray.map .ofName)
+    add_line $ m!"Found {hits.size} definitions mentioning {needles_list}."
+    unless (args.name_pats.isEmpty) do
+      let name_list := MessageData.andList <| args.name_pats.map fun s => m!"\"{s}\""
+      add_line $ m!"Of these, {hits2.size} have a name containing {name_list}."
+    unless (pats.isEmpty) do
+      add_line $ m!"Of these, {hits3.size} match your patterns."
+    unless (hits4.size ≤ maxShown) do
+      add_line $ m!"Of these, the first {maxShown} are shown."
+    return .inr (← summary.get, hits4.toList.take maxShown)
+
+/-!
+## The #find command, syntax and parsing
+-/
+
+open Lean.Parser
+
+/-- `#find` name pattern: `"substring"` -/
+syntax name_pattern := strLit
+/-- `#find` identifier pattern: `Real.sqrt`, `sqrt` -/
+syntax ident_pattern := ident
+/-- The turnstyle for `conclusion_pattern` -/
+syntax turnstyle := "⊢ " <|> "|- "
+/-- `#find` conclusion pattern: `⊢ (tsum _ = _)` -/
+syntax conclusion_pattern := turnstyle term:max
+/-- `#find` subexpression pattern: `(_ * (_ + _))` -/
+syntax term_pattern := term:max
+/-- `#find` patterns -/
+syntax find_pattern := name_pattern <|> ident_pattern <|> conclusion_pattern <|> term_pattern
+
+
+/-- Parses a list of `find_pattern` syntax into `Arguments` -/
+def parseFindPatterns (args : TSyntaxArray ``find_pattern) : TermElabM Arguments := do
+  let mut idents := #[]
+  let mut name_pats := #[]
+  let mut terms := #[]
+  for arg in args do
+    match arg with
+    | `(find_pattern| $ss:str) => do
+      let str := Lean.TSyntax.getString ss
+      name_pats := name_pats.push str
+    | `(find_pattern| $i:ident) => do
+      let n := Lean.TSyntax.getId i
+      unless (← getEnv).contains n do
+        throwErrorAt i "unknown identifier '{n}'"
+      idents := idents.push n
+    | `(find_pattern| $_:turnstyle $s:term) => do
+      let t ← Lean.Elab.Term.elabTerm s none
+      terms := terms.push (true, t)
+    | `(find_pattern| $s:term) => do
+      let t ← Lean.Elab.Term.elabTerm s none
+      terms := terms.push (false, t)
+    | _ => throwErrorAt arg "unexpected argument to #find"
+  pure {idents, name_pats, terms}
+
+open Lean.Elab.Command
+
+/--
+The `#find` command finds definitions and lemmas by various ways, which can be combined:
```

Comment:
```suggestion
The `#find_theorems` command finds definitions and lemmas by various ways, which can be combined:
```
Are both spellings `find` and `find_theorems` available?

Reply:
Ah, no. `find_theorems` is just temporary for testing until I actually move this over and replace  `#find`. Maybe this is more confusing than helpful :-)

Reply:
I wonder whether, for the PR, you should already assume that merging it means that the result will be an extension of `#find` and `#find_theorems` will not appear in mathlib (nor in its history).

Reply:
Exactly that!

This is all too confusing, I fear. I’ll just do that now, and if people really want to run the old `#find` while reviewing, they probably can switch checkouts easily.